engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/internal/PairwiseFloatRingBuffer.java

```diff
@@ -0,0 +1,381 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.engine.table.impl.updateby.internal;
+
+import gnu.trove.list.array.TIntArrayList;
+import io.deephaven.chunk.WritableFloatChunk;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.util.SafeCloseable;
+
+import java.util.NoSuchElementException;
+
+/***
+ * Store this data in the form of a binary tree where the latter half of the chunk is treated as a ring buffer and
+ * pairwise results of the `FloatFunction` are stored in the parent nodes.  We do lazy evaluation by maintaining a
+ * 'dirty' index list and computing the ultimate pairwise result only when requested by `evaluate()'
+ *
+ * To keep the parent-node finding math easy and consistent between the ring buffer and the computation tree, the binary
+ * tree is shifted by one index so the root (and final result of computation) ends up in index 1 (instead of 0 which is
+ * un-used)
+ */
+
+public class PairwiseFloatRingBuffer implements SafeCloseable {
+    // use a sized float chunk for underlying storage
+    private WritableFloatChunk<Values> storageChunk;
+    private final TIntArrayList dirtyIndices;
```

Comment:
We should think about this data type. `clear()` allocates a new array every time. Maybe we want to use row sets?

In discussing, we concluded that we have at most 1 range of pushed leaves and 1 range of popped leaves, and that we could instead accumulate those ranges, allowing for at most 3 calls to evaluateRangeFast per level. Then we could cascade, instead of tracking dirty.